zellij-server/src/screen.rs

```diff
@@ -1530,6 +1535,77 @@ impl Screen {
         }
     }
 
+    pub fn move_active_tab_to_left(&mut self, client_id: ClientId) -> Result<()> {
+        if self.tabs.len() < 2 {
+            info!("cannot move tab to left: only one tab exists");
+            return Ok(());
+        }
+        let Some(client_id) = self.client_id(client_id) else { return Ok(()); };
+        let Some(&active_tab_idx) = self.active_tab_indices.get(&client_id) else { return Ok(()); };
+        let left_tab_idx = self.index_on_left_of(active_tab_idx);
+        self.switch_tabs(active_tab_idx, left_tab_idx, client_id);
+        self.log_and_report_session_state()
+            .context("failed to move tab to left")?;
+        Ok(())
+    }
+
+    fn client_id(&mut self, client_id: ClientId) -> Option<u16> {
+        if self.get_active_tab(client_id).is_ok() {
+            Some(client_id)
+        } else {
+            self.get_first_client_id()
+        }
+    }
+
+    fn index_on_left_of(&self, current_index: usize) -> usize {
+        // wraps around: [tab1, tab2, tab3] => [tab1, tab2, tab3]
+        //                 ^                                 ^
+        //            current_index               index_on_left_of(current_index)
+        let tabs_count = self.tabs.len();
+        (current_index + tabs_count - 1) % tabs_count
+    }
+
+    fn switch_tabs(&mut self, active_tab_idx: usize, other_tab_idx: usize, client_id: u16) {
+        let (Some(mut active_tab), Some(mut other_tab)) = (
+            self.tabs.remove(&active_tab_idx),
+            self.tabs.remove(&other_tab_idx),
+        ) else {
```

Comment:
I have a question in here.

For unknown reasons, I think if `active_tab` is `Some` and `other_tab` is `None`, this condition is not satisfied and the function just returns.

However, since `remove` has already been called, `active_tab` will disappear from tabs.

For safer code, it seems better to check whether the key exists first. What do you think?

Reply:
That's a good catch. In this case it should be guaranteed that both `active_tab` and `other_tab` is `Some`, because the method `switch_tabs` is called only when:

1. There are at least two tabs.
2. `active_tab` is `Some`.

So only the `other_tab` thing left and it's guaranteed to be `Some` as well, because that's how `index_on_left_of` and `index_on_right_of` are implemented.

But anyway - who knows how the code will evolve, so I agree that more defensive approach is better. PR updated :+1: